tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/tee.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <optee_msg_supplicant.h>
+#include <tee_client_api.h>
+#include <teec_ta_load.h>
+
+LOG_MODULE_REGISTER(tee_supplicant);
+
+#define TEE_SUPP_THREAD_PRIO	7
+
+#define TEE_REQ_PARAM_MAX	5
+
+static struct k_thread main_thread;
+static K_THREAD_STACK_DEFINE(main_stack, 8192);
+
+struct suppl_shm {
+	int id;
+	void *p;
+	size_t size;
+	bool registered;
+	struct suppl_shm *next;
+};
+
+static K_MUTEX_DEFINE(shm_mutex);
+static struct suppl_shm *shm_head;
+
+static void tee_supp_mutex_lock(struct k_mutex *mu);
+static void tee_supp_mutex_unlock(struct k_mutex *mu);
+
+#define MEMREF_SHM_ID(p)	((p)->c)
+#define MEMREF_SHM_OFFS(p)	((p)->a)
+#define MEMREF_SIZE(p)		((p)->b)
+
+struct tee_supp_request {
+	uint32_t cmd;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+struct tee_supp_response {
+	uint32_t ret;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+union tee_supp_msg {
+	struct tee_supp_request req;
+	struct tee_supp_response rsp;
+};
+
+struct param_value {
+	uint64_t a;
+	uint64_t b;
+	uint64_t c;
+};
+
+static int receive_request(const struct device *dev, struct tee_supp_request *ts_req)
+{
+	int rc = tee_suppl_recv(dev, &ts_req->cmd, &ts_req->num_param, ts_req->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant receive failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int send_response(const struct device *dev, struct tee_supp_response *rsp)
+{
+	int rc = tee_suppl_send(dev, rsp->ret, rsp->num_param, rsp->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant send response failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int get_value(size_t num_params, struct tee_param *params,
+		     const uint32_t idx, struct param_value **value)
+{
+	if (idx >= num_params) {
+		return -1;
```

Comment:
Could you please define error codes somewhere? Also, is there are reason why standard TEE return codes are note used?

Reply:
This function is completely internal. Return code from this function never passed to outside function.

Reply:
You are using values `-1`, `-2`, `-3` as error codes in those internal functions. Why there are 3 of them and what they mean?

Reply:
It just for debugging. @oleksiimoisieiev insisted to this.

Reply:
Probably you've misunderstood my recommendation. What I was asking is why all return codes are -1 and suggested to use something meaningful. For example error codes from errno.h could be used. One of my arguments was that when you get correct return codes it could help for debugging purposes. Even if this function is only for the internal usage.
My vision is that the best thing will be to use TEEC_* error codes, so you can return them outside without updating. But if it's impossible then at least errno.h can be used